### PR TITLE
U/juliaroldi/table align

### PR DESCRIPTION
**Pull request by juliaroldi**
Align table using setAlignment API. 

tableAlignStyle

### CODE REVIEW

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 1]
```diff
@@ -3,34 +3,127 @@ import {
...
+ * @param addUndoSnapshot check if this function is being called by addUndoSnapshot
+ */
+function alignElement(
+    editor: IEditor,
+    element: HTMLElement,
+    elementType: string,
+    alignment: Alignment,
+    addUndoSnapshot?: boolean
+) {
+    if (elementType === TABLE && editor.isFeatureEnabled(ExperimentalFeatures.TableAlignment)) {
```
Comment (BryanValverdeU) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts]: We can check whether the whole table is selected using the editor api getSelectionRangeEx().

```
const selection = editor.getSelectionRangeEx();
if (selection.type == SelectionRangeTypes.TableSelection) {
 const vTable = new VTable(selection.table)
 const { firstCell, lastCell } = selection.coordinates;
    const rowsLength = vTable.cells.length - 1;
    const colIndex = vTable.cells[rowsLength].length - 1;
```

    const firstX = firstCell.x;
    const firstY = firstCell.y;
    const lastX = lastCell.x;
    const lastY = lastCell.y;

    const selectedAllTable = firstX == 0 && firstY == 0 && lastX == colIndex && lastY == rowsLength;
```
 if(selectedAllTable) { alignAllTable() }
}
```

Comment (juliaroldi) [reply packages/roosterjs-editor-api/lib/format/setAlignment.ts]: I see, if I select only some cells, the editor.getElementAtCursor returns Table. I will fix that.

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 2]
```diff
@@ -1,36 +1,146 @@
...
+    if (selection && selectionType === SelectionRangeTypes.Normal) {
+        return TEXT;
+    } else if (
+        editor.isFeatureEnabled(ExperimentalFeatures.TableAlignment) &&
+        selection &&
+        selectionType === SelectionRangeTypes.TableSelection &&
+        isWholeTableSelected(selection)
+    ) {
+        return TABLE;
+    }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:55]: else?

You can simplify your function to be:

return (type == table && isFeatureEnabled && isWholeTableSelected);

Making it returning a boolean should be easier to use

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 3]
```diff
@@ -1,36 +1,146 @@
...
+    elementType: string,
+    alignment: Alignment,
+    addUndoSnapshot?: boolean
+) {
+    if (elementType === TABLE) {
+        alignTable(editor, element, alignment);
+    } else {
+        alignText(editor, element, alignment, addUndoSnapshot);
+    }
+}
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:78]: I think all 3 functions above can be merged together:

```ts
function setAlignment(...) {
  if (type == table && isFeatureEnabled) {
    alignTable(..);
  } else {
    alignText(...);
}
```

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 4]
```diff
@@ -1,36 +1,146 @@
...
 export default function setAlignment(editor: IEditor, alignment: Alignment) {
-    let command = DocumentCommand.JustifyLeft;
-    let align = 'left';
+    const element = editor.getElementAtCursor();
+    const elementType = isATableOrText(element, editor);
+
+    editor.addUndoSnapshot(() => {
+        alignElement(editor, element, elementType, alignment);
+        editor.queryElements('[align]', QueryScope.OnSelection, node =>
+            alignElement(editor, node, elementType, alignment, true /** addUndoSnapshot */)
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:30]: why we need to call alignElement twice in this funciton?

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 5]
```diff
@@ -1,36 +1,146 @@
...
         command = DocumentCommand.JustifyCenter;
         align = 'center';
     } else if (alignment == Alignment.Right) {
         command = DocumentCommand.JustifyRight;
         align = 'right';
     }
-
-    editor.addUndoSnapshot(() => {
+    if (addUndoSnapshot) {
+        element.style.textAlign = align;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:125]: why "addUndosnapshot" can determin if we want to set textAlign?

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 6]
```diff
@@ -1,36 +1,146 @@
...
-    }, ChangeSource.Format);
+    }
+}
+
+/**
+ * Check if the whole table is selected
+ * @param selection
+ * @returns
+ */
+function isWholeTableSelected(selection: TableSelectionRange) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:136]: what if I select more than whole table?

Comment (juliaroldi) [reply packages/roosterjs-editor-api/lib/format/setAlignment.ts:84]: Only the text is aligned. 
tableAndText

Comment (JiuqingSong) [reply packages/roosterjs-editor-api/lib/format/setAlignment.ts:84]: I think we should center-align both text outside the table and the table itself

Comment (JiuqingSong) [reply packages/roosterjs-editor-api/lib/format/setAlignment.ts:84]: Ok, I checked the behavior or Word, it is the same with your current implementation. So I think it is fine here.

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 7]
```diff
@@ -14,23 +18,77 @@ import {
...
 export default function setAlignment(editor: IEditor, alignment: Alignment) {
-    let command = DocumentCommand.JustifyLeft;
-    let align = 'left';
+    const element = editor.getElementAtCursor();
+    const selection = editor.getSelectionRangeEx();
+    const selectionType = selection.type;
+    editor.addUndoSnapshot(() => {
+        if (
+            editor.isFeatureEnabled(ExperimentalFeatures.TableAlignment) &&
+            selection &&
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:27]: this need to be moved before line 23. you already used selection before check it

File: packages/roosterjs-editor-api/lib/format/setAlignment.ts [hunk 8]
```diff
@@ -14,23 +18,77 @@ import {
...
 
+/**
+ * Align text using the margins
+ * @param editor
+ * @param element
+ * @param alignment
+ * @param addUndoSnapshot
+ * @returns
+ */
+function alignTable(editor: IEditor, element: HTMLElement, alignment: Alignment) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/setAlignment.ts:46]: editor is not used

**Review by BryanValverdeU — APPROVED** [review]

**Review by JiuqingSong — APPROVED** [review]

**Pull request merged**